[rtloader] Enable dynamic interpreter binary path resolution

sgnn7 opened the pull request:
### What does this PR do?

Previously, our RTLoader was unable to have a custom program name (via
`Py_SetProgramName`) which made use of libraries like `multiprocessing`
not work since they would try to spawn new processes via the rtloader's
parent process path (in this case the agent binary).

**Caveats: Processes spawned from the main one are not injected with the Golang helpers from
the Agent (eg. `from datadog_agent import ...`). Due to this, all interactions with said helpers
must be done from the main process and subprocesses must use `try`/`catch` guards around
Golang helper imports.**

### Motivation

This change adds an ability to configure this value so that when
a library like `multiprocessing` it can point to the correct interpreter
path and work as intended.

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

- Ensure that Python integrations work as normal
- Ensure that `sys.executable` within a check shows an absolute path to the python interpreter binary and not the agent binary
- Alternatively, you can ensure that checks using libraries like `multiprocessing` with `spawn` do not error out on the main 3 platforms (example check can be found [here](https://gist.github.com/sgnn7/8e3a64f27e0770ae580ebc8b3cf13c02).

sgnn7 committed:
[rtloader] Add ability to set program name to rtloader

Previously, our RTLoader was unable to have a custom program name (via
`Py_SetProgramName`) which made use of libraries like `multiprocessing`
not work since they would try to spawn new processes via the rtloader's
pareent process path.

This change adds an ability to configure this value so that when
a library like `multiprocessing` it can point to the correct interpreter
path and work as intended.

Signed-off-by: Srdjan Grubor <sgnn7@sgnn7.org>

sgnn7 committed:
[rtloader] Fixup some of the issues with the demo program

The demo program seems like it was abandoned to some extent as the paths
did not work and the callback signature was out of date. This change
allows it to be buildable and possibly runnable though the demo code has
an unrelated segmentation fault when run from Darwin that would need
futher looking into.

Signed-off-by: Srdjan Grubor <sgnn7@sgnn7.org>

sgnn7 committed:
[rtloader] Launch rtloader with inferred Python interpreter's path

Now that RTLoader can accept the path to the Python interpreter as a
parameter, this change sets this value to the inferred one from the
ldflags and the environment.

Signed-off-by: Srdjan Grubor <sgnn7@sgnn7.org>

sgnn7 committed:
[rtloader] Use upstream workaround for setting Python bin path on Windows

Due to an [upstream bug](https://bugs.python.org/issue34725) in how
Windows handles `Py_SetProgramName`, we have to use the suggested
workaround and also use `_Py_SetProgramFullPath` for now to make this
code work on that platform.

Since this is a workaround API in CPython, it is depricated as of
Python 3.11 and will be removed at some later version. Before it gets
removed, we should convert our Python 3 initialization code to the
more-appropriate `PyConfig API`
(https://docs.python.org/3.11/c-api/init_config.html#c.PyConfig).

ogaca-dd commented on rtloader/two/two.cpp:
```diff
@@ -52,16 +59,33 @@ Two::~Two()
+
+    _pythonExe = strdupe(python_exe);
+
+    Py_SetProgramName(_pythonExe);
```
Based from https://docs.python.org/3/c-api/init.html#c.Py_SetProgramName, `Py_SetProgramName` should take as argument `const wchar_t *`.

sgnn7 replied:
The docs you linked are for Python 3 but this is the Python 2 impl - [it takes a `char *`](https://docs.python.org/2/c-api/init.html#c.Py_SetProgramName)

ogaca-dd commented on rtloader/demo/main.c:
```diff
@@ -26,7 +28,7 @@ char **get_tags(char *id, int highCard)
 }
 
-void submitMetric(char *id, metric_type_t mt, char *name, double val, char **tags,  char *hostname)
+void submitMetric(char *id, metric_type_t mt, char *name, double val, char **tags,  char *hostname, bool flush_first_val)
```
`flush_first_val` is unused.

sgnn7 replied:
true but it is required in the callback otherwise it doesn't compile

ogaca-dd replied:
I assumed the code compile before your changes and you just added a new parameter!

sgnn7 commented on rtloader/two/two.cpp:
```diff
@@ -52,16 +59,33 @@ Two::~Two()
-    // Py_SetPythonHome stores a pointer to the string we pass to it, so we must keep it in memory
     Py_SetPythonHome(_pythonHome);
-    _free(oldPythonHome);
+    if (oldPythonHome) {
```
Initially, the value of the `oldPythonHome` is `NULL` so trying to free it should be a problem. This check ensures that we only free the old value if it has been set previously. I know we [check for NULL in `_free` itself](https://github.com/DataDog/datadog-agent/blob/140a4ee164261ef2245340c50371ba989fbeb038/rtloader/common/rtloader_mem.c#L50) but that seems to be an implementation detail that we shouldn't rely on.

Would love your thoughts on this though.

ogaca-dd replied:
I think every custom implementation of `free` must handle `NULL` ([POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)) but it is also 100% OK to assume it is not the case! I wanted to make sure, it was not a temporary line committed!

sgnn7 committed:
[ci] Fix issues with upgrading appveyor image

With the newer image that appveyor uses, some tests are running slower
and failing. This change fixes some tests to be more resiliant to this
change.

truthbk commented on rtloader/test/rtloader/rtloader_test.go:
```diff
@@ -69,6 +69,28 @@ with open(r'%s', 'w') as f:
 	helpers.AssertMemoryUsage(t)
 }
 
+func TestSysExecutableValue(t *testing.T) {
```
Nice test 👌

sgnn7 replied:
👍 it's what found the problem with the Windows `Py_SetProgramName` impl :)

truthbk commented on rtloader/three/three.cpp:
```diff
@@ -50,18 +57,36 @@ Three::~Three()
+
+    Py_SetProgramName(_pythonExe);
+
+    // HACK: This extra internal API invocation is due to the workaround for an upstream bug on
```
If it's just necessary on windows can we put that line within a preprocessor block?
```
#if defined(_WIN32)
  _Py_SetProgramFullPath(_pythonExe);
#endif
```
I believe this would be accurate but you may want to check with someone more current and knowledgeable on all-things-windows than me 😊

sgnn7 replied:
I feel a bit conflicted on this one. While we can easily do what you are saying, there are a few upsides to having it without a guard `define` that made me choose otherwise:
- Consistent codebase for all platforms
- It's an effective noop on other platforms afaik since the method above it should be doing the same exact thing. 
- Tests pass fine as-is on all used platforms.
- Problems can be found quicker with building the code as it won't just fail on Windows.
- This function is deprecated w/ 3.11+ so when we start seeing warnings for that we will have it on all the platforms as a good reminder that we should move to the new way for Py configuration.

Overall, the benefits of not platform-guarding that call to me seemed to outweigh the downsides. I'd love your thoughts on it though.

truthbk replied:
OK, I'm fine with this.

truthbk reviewed (commented):
Really nice stuff, added some comments/questions! This is useful! Thank you :)

sgnn7 commented on appveyor.yml:
```diff
@@ -5,6 +5,8 @@ skip_branch_with_pr: true
 
 clone_folder: C:\gopath\src\github.com\DataDog\datadog-agent
 
+image: "Visual Studio 2017"
```
Yup! This change was made w/ consultation with @KSerrania (you can look for `I tried upgrading Python from 3.7 to 3.8` on slack for the full thread.

sgnn7 commented on pkg/collector/python/init.go:
```diff
@@ -164,6 +165,8 @@ void initkubeutilModule(rtloader_t *rtloader) {
 */
 import "C"
 
+const PythonWinExeBasename = "python.exe"
```
I agree. I think this module would ideally do good with a split-out of some things into the platform-specific modules like you say but at current time it would have been a much bigger change than desired, especially since we play with module-level vars in the affected methods. I can keep that in mind for future work on this. Thoughts?

truthbk replied:
Yeah, no worries, I don't think you need to do this here. We should maybe create a card to try and remember to do that. It may take some time as it'd just be a refactor, but probably still worth it.

sgnn7 commented on pkg/collector/python/init.go:
```diff
@@ -278,48 +293,89 @@ func detectPythonLocation(pythonVersion string) {
+		interpreterAbsPath, err := resolveExecPath(interpreterBasename)
+		if err != nil {
+			log.Warnf("Error trying to resolve interpreter path for executable: '%v'", err)
+			return interpreterBasename
```
A fatal error here would be inappropriate I think since almost everything would still work. The Python integrations would all still work it's just that the edge cases (like `multiprocessing` lib) would fail to run under `spawn` settings.

truthbk replied:
If we're going to advertise `multiprocessing` as supported, then it does feel like it could be considered a fatal error to me.... At the very least we should then log something about what wouldn't work.

sgnn7 commented on rtloader/demo/CMakeLists.txt:
```diff
@@ -1,11 +1,40 @@
+  set_target_properties(demo PROPERTIES LINK_FLAGS -static)
+  target_link_libraries(demo PUBLIC datadog-agent-rtloader)
+else()
+  if (APPLE)
```
Hmm my CMake knowledge is minimal but I thought you needed to specify the full filename (including the extension) if you're linking to a non-system lib. I'll double check though.

sgnn7 replied:
Looks like you need the extension when using abs paths for libraries:
```
...
make[3]: *** No rule to make target `../build/rtloader/libdatadog-agent-rtloader', needed by `demo'.  Stop.
...
```

truthbk replied:
I'm really surprised this wasn't building properly without that directive, it's harmless if it works but CMake would expectedly automatically handle it. Anyways, no need to waste more time on it, especially not on the demo CMake.

sgnn7 replied:
I believe that when using just the system library name it's smart enough to figure it out but when using absolute path, it needs to know the full thing. The original code was assuming that you have the system library installed but even when you do `inv rtloader.install` it doesn't really make it available to the demo on non-*nix platforms as I found out (thus the workaround with absolute paths).

sgnn7 commented on rtloader/demo/CMakeLists.txt:
```diff
@@ -1,11 +1,40 @@
+
+## Add our paths for includes. They may be different on the directory we are
+## starting our build from
+if (EXISTS "${CMAKE_SOURCE_DIR}/include")
```
If you're trying to build the demo from this repo, the include files are not in `<demo_dir>/include` but in the parent (`<demo_dir>/../include` and common). Given that I didn't quite know how the old values (`<demo_dir>/include`) were used and because there's no downsides to including an `if`/`else` I left in the fallback. If you think the original code was bad, I'll just remove the L14 `true` branch.

truthbk replied:
I missed the fact this was the `CMakeLists.txt` for the demo, not the main project. 🤦 

Answering your question, we probably don't need the original code, just your fix. There's nothing wrong with cleaning up what was wrong :)

Probably what happened is we shuffled/de-duped some code and never really verified this was still building. I'm all 👍 for removing the old broken stuff.

albertvaka reviewed (approved):
👍 for changes to appveyor.yml and invoke tasks.

sgnn7 committed:
[rtloader] Add `allow_python_path_heuristics_failure` flag to config

This flag can now decide if we will continue loading the rtloader or not
in cases where we can't reliably deduce the Python interpreter's full path.

truthbk commented on pkg/collector/python/init.go:
```diff
@@ -298,20 +298,26 @@ func resolvePythonExecPath(pythonVersion string) string {
 					" Python's 'multiprocessing' library may fail to work.", err)
+				return PythonWinExeBasename, nil
+			} else {
+				return "", fmt.Errorf("Error trying to resolve interpreter path for executable: '%v'."+
```
I think currently '%v' going to be the `err` value not the executable name, so this error might look a little weird.

Also, since we're using pretty much the same error message in two places and for consistency and maintainability, we could maybe implement a custom error. Feel free to ignore, just a nit.

sgnn7 replied:
[Fixed](https://github.com/DataDog/datadog-agent/pull/8369/commits/e10fde45b48576c9e71f8dfc8aa0decba744f82a)

truthbk commented on pkg/config/config.go:
```diff
@@ -247,6 +247,12 @@ func InitConfig(config Config) {
+	// is expected to prevent rtloader from initializing. When set to true, this override
+	// allows us to proceed but with some capabilities unavailable (e.g. `multiprocessing`
+	// library support will not work reliably in those environments)
+	config.BindEnvAndSetDefault("allow_python_path_heuristics_failure", false)
```
I think this is indeed the correct default.

sgnn7 replied:
The full path config setting approach I think was deferred until a later time

truthbk reviewed (commented):
Looks pretty good to me!! :)

xornivore commented on pkg/process/config/config_test.go:
```diff
@@ -20,19 +21,77 @@ import (
 	"google.golang.org/grpc"
 )
 
+type fakeContainerImpl struct {
```
I could see this fake noop implementation useful in other scenarios and for tests - maybe move it to `github.com/DataDog/datadog-agent/pkg/util/containers/providers` ?

sgnn7 replied:
@xornivore [Fixed](https://github.com/DataDog/datadog-agent/pull/8369/commits/c2cef0707f39dbf5e4c26c8a9aa8001366590ab0)

sgnn7 commented on rtloader/demo/main.c:
```diff
@@ -78,15 +80,15 @@ int main(int argc, char *argv[])
     // Embed Python2
     if (strcmp(argv[1], "2") == 0) {
-        rtloader = make2(python_home, &init_error);
+        rtloader = make2(python_home, "python2", &init_error);
```
It can be but the demo doesn't use this value at all so I used this just as the placeholder. I'll just swap it with an empty string to not cause confusion.

sgnn7 commented on pkg/collector/python/init.go:
```diff
@@ -278,48 +310,112 @@ func detectPythonLocation(pythonVersion string) {
 
 func Initialize(paths ...string) error {
 	pythonVersion := config.Datadog.GetString("python_version")
+	allowPathHeuristicsFailure := config.Datadog.GetBool("allow_python_path_heuristics_failure")
```
Moved function to more appropriate place since it's a generic func (`pkg/util/executable`) so we probably won't need to this anymore

sgnn7 commented on rtloader/include/datadog_agent_rtloader.h:
```diff
@@ -29,24 +29,26 @@ struct rtloader_pyobject_s;
 
 // FACTORIES
-/*! \fn rtloader_t *make2(const char *python_home, char **error)
+/*! \fn rtloader_t *make2(const char *python_home, const char *python_exe, char **error)
```
Yeah I thought about this option for a while too but there's a problem of consistency since:
-This can easily be confused with the python interpreter's `bin` folder that holds the executable instead of the abs path to the interpreter binary itself.
- `exec`/`executable`([link](https://docs.python.org/3/c-api/init.html#c.Py_GetExecPrefix), [link](https://docs.python.org/3/c-api/init.html#c.Py_SetPath)) is the terminology used in the Python C API to indicate the path to the Python interpreter.

So I thought the lower chance of confusion would be to apply the C API common pattern. Maybe we can do `python_exe` -> `python_exec` instead? I don't have a strong opinion either way though so up to you how you would like this fixed up.

hush-hush replied:
You'r right, `python_exe` is better then.

hush-hush commented on releasenotes/notes/set-rtloader-python-interpreter-bin-path-da017336fd301733.yaml:
```diff
@@ -0,0 +1,9 @@
+    Python interpreter ``sys.executable`` is now set to the appropriate interpreter's
+    executable path. This should allow ``multiprocessing`` to be able to spawn new
+    processes since it will try to invoke the Python interpreter instead of the Agent
+    itself. It should be noted though that the ``datadog_agent`` import should be
+    guarded with ``try``/``except`` clauses and only used from within the main process
+    as the ``rtloader``-injected helpers are not available in the child processes.
```
I don't think we should advertise way to handle the import. WDYT about:

```suggestion
    itself. It should be noted though that the Pyton packages injected at runtime by the Agent are only available from the main process, not from any sub-processes.
```

sgnn7 committed:
[docs] Update releasenotes/notes/set-rtloader-python-interpreter-bin-path-da017336fd301733.yaml

Co-authored-by: maxime mouial <hush-hush@users.noreply.github.com>

sgnn7 committed:
[rtloader] Use basename of interpreter if ignoring errors

If we are ignoring errors, setting the interpreter name to the basename
of the expected executable is more likely to work than the alternatives
as we will always fail to find the interpreter otherwise.
